torch/csrc/dynamo/eval_frame.c

```diff
@@ -327,24 +419,28 @@ static CacheEntry* create_cache_entry(
   //   - guarded_code: Borrowed
   //  return
   //   - CacheEntry*: new reference.
-  CacheEntry* e = (CacheEntry*)malloc(sizeof(CacheEntry));
-  DEBUG_NULL_CHECK(e);
-  e->check_fn = PyObject_GetAttrString(guarded_code, "check_fn");
-  NULL_CHECK(e->check_fn);
-  e->code = (PyCodeObject*)PyObject_GetAttrString(guarded_code, "code");
-  NULL_CHECK(e->code);
-  e->next = next;
+  PyObject* check_fn = PyObject_GetAttrString(guarded_code, "check_fn"); // new reference
+  PyCodeObject* code = (PyCodeObject*)PyObject_GetAttrString(guarded_code, "code"); // new reference
+
+  // equivalent to CacheEntry(check_fn, code, next) in Python
+  PyObject* args = Py_BuildValue("OOO", check_fn, code, next);
+  CacheEntry* e = (CacheEntry*)PyObject_CallObject((PyObject*)&CacheEntryType, args); // new reference
+  // CacheEntry e is the now the owner of old cachey entry next. This happens
+  // when we incref the next pointer in cache_entry_init.
+  Py_DECREF(next);
+  Py_DECREF(check_fn);
+  Py_DECREF(code);
+  Py_DECREF(args);
```

Comment:
@ezyang One last review of these lines - https://github.com/pytorch/pytorch/pull/107405/files#r1299639440